### ISSUE TITLE
SHL & ROL constant folding inaccuracy

### ISSUE THREAD

**Issue by ek0** [Vector35/binaryninja-api#2869]
**Version and Platform (required):**
- Binary Ninja Version: Version 3.0.3208-dev Personal
- OS: Windows 11
- OS Version: 21H2

**Bug Description:**
Given these two function:
```x86asm
test_shl_maxfold PROC
	mov rcx, 08000000000000000h
	shl rcx, 1
	mov rax, rcx
	ret
test_shl_maxfold ENDP

test_rol_maxfold PROC
	mov rcx, 08000000000000000h
	rol rcx, 1
	mov rax, rcx
	ret
test_rol_maxfold ENDP
```

Their respective results are:
```
PS D:\codes\ConsoleApplication2\x64\Debug> .\adder.exe
test_rol_maxfold 1
test_shl_maxfold 0
```

But in HLIL view, we get these results:
![image](https://user-images.githubusercontent.com/7302683/150713756-44f454ec-605e-44ed-9faf-4513fef2955d.png)
![image](https://user-images.githubusercontent.com/7302683/150713776-bf94a122-5180-4fff-b3ef-f81628cb42bb.png)

**Comment by ek0** [comment]
Just a note since I haven't seen an update. SHL works fine, wrong folding was due to a patch for the lifting and has been fixed in the appropriate [branch](https://github.com/Vector35/arch-x86/pull/28). ROL folding is buggy though.

**Comment by psifertex** [comment]
Sorry for the delay in responding here -- we're triaging ROL folding internally now and will update the tags appropriately when it's done.

**Comment by bpotchik** [comment]
Fixed in `3.0. 3307-dev`.
